File: backend/agents/sim_recommender_agent.py
"""
SIM Recommender Agent
Picks the best provider and explains why.
Deterministic — no LLM required. Can be enhanced with LFM if available.
"""
from __future__ import annotations

from typing import Any

from ..services.recommendation_service import (
    choose_best_provider,
    generate_recommendation_text,
    summarize_reports,
)


class SimRecommenderAgent:
    """Recommends the best SIM/provider for the analyzed location or route."""

    def run(self, analysis_result: dict[str, Any]) -> dict[str, Any]:
        provider_scores = analysis_result.get("provider_scores", {})
        weak_segments = analysis_result.get("weak_segments", [])
        nearby_reports = analysis_result.get("nearby_reports", [])
        report_summary = summarize_reports(nearby_reports)

        best_provider = choose_best_provider(provider_scores)
        score_data = provider_scores.get(best_provider, {})
        score_value = score_data.get("score", 0) if isinstance(score_data, dict) else 0

        # Build ranked list
        ranked = sorted(
            [
                {
                    "provider": name,
                    "score": round(float(data.get("score", 0)), 1),
                    "coverage_pct": round(float(data.get("coverage_rate_percent", 0)), 1),
                }
                for name, data in provider_scores.items()
                if isinstance(data, dict)
            ],
            key=lambda x: x["score"],
            reverse=True,
        )

        # Determine confidence
        if score_value >= 80:
            confidence = "high"
        elif score_value >= 50:
            confidence = "medium"
        else:
            confidence = "low"

        # Build caveats
        caveats: list[str] = []
        if weak_segments:
            caveats.append(f"{len(weak_segments)} weak signal segment(s) detected along the route.")
        if report_summary.get("total_reports", 0) > 0:
            common = report_summary.get("common_issue")
            if common and common not in ("unknown", "other"):
                caveats.append(f"Community reports indicate '{common}' issues in this area.")
        if score_value < 50:
            caveats.append("Coverage data is sparse — treat recommendation with caution.")

        return {
            "recommended_provider": best_provider,
            "score": round(float(score_value), 1),
            "summary": generate_recommendation_text(provider_scores, weak_segments),
            "confidence": confidence,
            "ranked_providers": ranked,
            "caveats": caveats,
            "nearby_reports_summary": report_summary,
        }

File: frontend/src/agents/simRecommenderAgent.ts
import { LFMClient } from "../ai/lfmClient";
import { buildJsonAgentPrompt } from "../ai/prompts";

export type SimRecommendationOutput = {
  recommended_provider: string;
  summary: string;
  confidence: "low" | "medium" | "high";
  caveats: string[];
};

export class SimRecommenderAgent {
  constructor(private lfm: LFMClient) {}

  async run(analysisResult: any): Promise<SimRecommendationOutput> {
    const fallback: SimRecommendationOutput = {
      recommended_provider: analysisResult?.best_provider || "Unknown",
      summary:
        analysisResult?.recommendation_text ||
        "No provider recommendation is available.",
      confidence:
        analysisResult?.weak_segments?.length > 0
          ? "medium"
          : analysisResult?.best_provider
            ? "high"
            : "low",
      caveats: analysisResult?.offline_readiness_alerts || [],
    };

    try {
      return await this.lfm.generateJson<SimRecommendationOutput>(
        buildJsonAgentPrompt({
          agentName: "Sim Recommender Agent",
          task:
            "Recommend the best provider based on provider_scores, weak segments, and report summary.",
          input: {
            best_provider: analysisResult?.best_provider,
            provider_scores: analysisResult?.provider_scores,
            weak_segments: analysisResult?.weak_segments,
            nearby_reports_summary: analysisResult?.nearby_reports_summary,
            recommendation_text: analysisResult?.recommendation_text,
          },
          outputShape: fallback,
        }),
        512
      );
    } catch {
      return fallback;
    }
  }
}